### PR TITLE
refine pybind when *_op.cc contains several operators

### PR DESCRIPTION
**Pull request by luotao1**
fix https://github.com/PaddlePaddle/Paddle/issues/7059#issuecomment-354111897
- for most op contains several operators, such as `conv_op`, `conv_cudnn_op`, etc
  - auto pybind it by `REGEX MATCH REGISTER_OP(***,`, where `***` is the name for `USE_OP`/`USE_CPU_ONLY_OP` etc .
- for other op whose `REGISTER_OP` is manually, such as `compare_op`:
https://github.com/PaddlePaddle/Paddle/blob/fba6a10dd99edf6110280754555af78889f19dd3/paddle/operators/compare_op.cc#L81-L105
   - manually pybind them.

### CODE REVIEW

File: paddle/operators/CMakeLists.txt [hunk 1]
```diff
@@ -61,106 +61,26 @@ function(op_library TARGET)
...
-        file(APPEND ${pybind_file} "USE_CUDA_ONLY_OP(ncclAllReduce);\n")
-    endif()
-
-    # reduce_op contains several operators
-    if ("${TARGET}" STREQUAL "reduce_op")
-        set(pybind_flag 1)
-        # It's enough to just adding one operator to pybind
-        file(APPEND ${pybind_file} "USE_OP(reduce_sum);\n")
-    endif()
+    # net_op doesn't need pybind, others will be pybind manually
```
Comment (typhoonzero) [inline paddle/operators/CMakeLists.txt]: This line of comment can be:

Define operators that don't need pybind here.

File: paddle/operators/CMakeLists.txt [hunk 2]
```diff
@@ -61,106 +61,26 @@ function(op_library TARGET)
...
 
-    if ("${TARGET}" STREQUAL "tensor_array_read_write_op")
-        set(pybind_flag 1)
-        file(APPEND ${pybind_file} "USE_NO_KERNEL_OP(read_from_array);\nUSE_NO_KERNEL_OP(write_to_array);\n")
+    file(READ ${TARGET}.cc TARGET_CONTENT)
+    # It's enough to just adding one operator to pybind
+    string(REGEX MATCH "REGISTER_OP\\(.*REGISTER_OP\\(" multi_register "${TARGET_CONTENT}")
+    string(REGEX MATCH "REGISTER_OP\\([a-z0-9_]*," one_register "${multi_register}")
+    if (one_register STREQUAL "")
+        string(REPLACE "_op" "" TARGET "${TARGET}")
```
Comment (typhoonzero) [inline paddle/operators/CMakeLists.txt]: Is it necessary to add some debug message here to see the registered operators?

File: paddle/operators/CMakeLists.txt (hunk 1 above)
Comment (luotao1) [reply paddle/operators/CMakeLists.txt]: Done. Thank you.

File: paddle/operators/CMakeLists.txt (hunk 2 above)
Comment (luotao1) [reply paddle/operators/CMakeLists.txt]: As there are hundreds of operators, the registered message are a lot too. If we want to see them, we can refer to `paddle/pybind/pybind.h`.

**Comment by tonyyang-svail** [comment]
hi @luotao1, just to confirm the reason for "[it's enough to just adding one operator to pybind in a *_op.cc file](https://github.com/PaddlePaddle/Paddle/blob/6f8314235327848906bd7800743be376b01cdd74/paddle/fluid/operators/CMakeLists.txt#L109-L111)" is that the linking of a single operator causes the linking of the whole object file, therefore all operators in that object file will be registered.

**Pull request merged**